src/daft-logical-plan/src/optimization/rules/reorder_joins/greedy_join_order.rs

```diff
@@ -0,0 +1,153 @@
+use std::{collections::HashMap, sync::Arc};
+
+use common_error::DaftResult;
+use daft_dsl::{col, ExprRef};
+
+use super::join_graph::{JoinCondition, JoinGraph};
+use crate::{LogicalPlanBuilder, LogicalPlanRef};
+
+// This is an implementation of the Greedy Operator Ordering algorithm (GOO) [1] for join selection. This algorithm
+// selects join edges greedily by picking the edge with the smallest cost at each step. This is similar to Kruskal's
+// minimum spanning tree algorithm, with the caveat that edge costs update at each step, due to changing cardinalities
+// and selectivities between join nodes.
+//
+// Compared to DP-based algorithms, GOO is not always optimal. However, GOO has a complexity of O(n^3) and is more viable
+// than DP-based algorithms when performing join ordering on many relations. DP Connected subgraph Complement Pairs (DPccp) [2]
+// is the DP-based algorithm widely used in database systems today and has a O(3^n) complexity, although the latest
+// literature does offer a super-polynomially faster DP-algorithm but that still has a O(2^n) to O(2^n * n^3) complexity [3].
+//
+// For this reason, we maintain a greedy-based join ordering algorithm to use when the number of relations is large, and default
+// to DP-based algorithms otherwise.
+//
+// [1]: Fegaras, L. (1998). A New Heuristic for Optimizing Large Queries. International Conference on Database and Expert Systems Applications.
+// [2]: Moerkotte, G., & Neumann, T. (2006). Analysis of two existing and one new dynamic programming algorithm for the generation of optimal bushy join trees without cross products. Very Large Data Bases Conference.
+// [3]: Stoian, M., & Kipf, A. (2024). DPconv: Super-Polynomially Faster Join Ordering. ArXiv, abs/2409.08013.
+pub(crate) struct GreedyJoinOrderer {}
+
+impl GreedyJoinOrderer {
+    /// Consumes the join graph and transforms it into a logical plan with joins reordered.
+    pub(crate) fn compute_join_order(join_graph: &mut JoinGraph) -> DaftResult<LogicalPlanRef> {
```

Comment:
I think this is doing too much (reordering and creating the logical plan) and isn't really that modular.

Instead what I would recommend is the following:

```
Trait JoinReorderer {
  fn reorder(&self, &mut graph: JoinGraph)
}

LogicalPlan -> JoinGraphBuilder -> JoinGraph -> JoinReorderer.reorder(JoinGraph) -> JoinGraph -> LogicalPlan
```

This way you get two things:
1. must easier to test that the JoinGraph roundtrip is working correctly
2. much easier to plug in a JoinReorderer (just implement that Trait)